crates/ruff_python_ast/generate.py

```diff
@@ -0,0 +1,752 @@
+#!/usr/bin/python
+# /// script
+# requires-python = ">=3.11"
+# dependencies = []
+# ///
+
+from __future__ import annotations
+
+import re
+from dataclasses import dataclass
+from pathlib import Path
+from subprocess import check_output
+from typing import Any
+
+import tomllib
+
+
+def rustfmt(code: str) -> str:
+    return check_output(["rustfmt", "--emit=stdout"], input=code, text=True)
+
+
+def to_snake_case(node: str) -> str:
+    """Converts CamelCase to snake_case"""
+    return re.sub("([A-Z])", r"_\1", node).lower().lstrip("_")
+
+
+# ------------------------------------------------------------------------------
+# Read AST description
+
+
+def load_ast(root: Path) -> list[Group]:
+    ast_path = root.joinpath("crates", "ruff_python_ast", "ast.toml")
+    with ast_path.open("rb") as ast_file:
+        ast = tomllib.load(ast_file)
+    return [Group(group_name, group) for group_name, group in ast.items()]
+
+
+# ------------------------------------------------------------------------------
+# Preprocess
+
+
+@dataclass
+class Group:
+    name: str
+    nodes: list[Node]
+    owned_enum_ty: str
+    ref_enum_ty: str
+
+    add_suffix_to_is_methods: bool
+    anynode_is_label: str
+    rustdoc: str | None
+
+    def __init__(self, group_name: str, group: dict[str, Any]) -> None:
+        self.name = group_name
+        self.owned_enum_ty = group_name
+        self.ref_enum_ty = group.get("ref_enum_ty", group_name + "Ref")
+        self.add_suffix_to_is_methods = group.get("add_suffix_to_is_methods", False)
+        self.anynode_is_label = group.get("anynode_is_label", to_snake_case(group_name))
+        self.rustdoc = group.get("rustdoc")
+        self.nodes = [
+            Node(self, node_name, node) for node_name, node in group["nodes"].items()
+        ]
+
+
+@dataclass
+class Node:
+    name: str
+    variant: str
+    ty: str
+
+    def __init__(self, group: Group, node_name: str, node: dict[str, Any]) -> None:
+        self.name = node_name
+        self.variant = node.get("variant", node_name.removeprefix(group.name))
+        self.ty = f"crate::{node_name}"
+
+
+# ------------------------------------------------------------------------------
+# Preamble
+
+
+def write_preamble(out: list[str]):
```

Comment:
nit: I would personally add return annotations to all functions, even those which just return `None`

```suggestion
def write_preamble(out: list[str]) -> None:
```

but I know some find this just needless boilerplate. No strong feeling! :-)

Reply:
Done!  (I must be too used to Rust, where that's the default :sweat_smile:)